PWG/EMCAL/EMCALtasks/AliEmcalCorrectionClusterNonLinearityMCAfterburner.cxx

```diff
@@ -0,0 +1,552 @@
+      if (fNonLinearityAfterburnerFunction != 0) {
+
+        //This is function Iteration-1 FunctionNL_kSDM combined with Iteration-2 FunctionNL_kSDM from AliCaloPhotonCuts
+        if (fNonLinearityAfterburnerFunction == 1) {
+        		newEnergy = oldEnergy;
+        		newEnergy/= fNLAfterburnerPara[0] + exp( fNLAfterburnerPara[1] + (fNLAfterburnerPara[2] * newEnergy));
+        		newEnergy/= fNLAfterburnerPara[3] + exp( fNLAfterburnerPara[4] + (fNLAfterburnerPara[5] * newEnergy));
+        }
+        //This is function FunctionNL_DPOW combined with FunctionNL_kSDM from AliCaloPhotonCuts
+        if (fNonLinearityAfterburnerFunction == 2) {
+        		newEnergy = oldEnergy;
+        		newEnergy/= (fNLAfterburnerPara[0] + fNLAfterburnerPara[1] * TMath::Power(oldEnergy, fNLAfterburnerPara[2]))/(fNLAfterburnerPara[3] + fNLAfterburnerPara[4] * TMath::Power(oldEnergy, fNLAfterburnerPara[5]));
+       		newEnergy/= fNLAfterburnerPara[6] + exp( fNLAfterburnerPara[7] + (fNLAfterburnerPara[8] * newEnergy));
+        }
+        //This is function FunctionNL_DExp combined with FunctionNL_kSDM from AliCaloPhotonCuts
+        if (fNonLinearityAfterburnerFunction == 3) {
+      			newEnergy = oldEnergy;
+      			newEnergy/= ((fNLAfterburnerPara[0] - TMath::Exp(-fNLAfterburnerPara[1]*oldEnergy +fNLAfterburnerPara[2]))/(fNLAfterburnerPara[3] - TMath::Exp(-fNLAfterburnerPara[4]*oldEnergy +fNLAfterburnerPara[5])));
+        		newEnergy/= fNLAfterburnerPara[6] + exp( fNLAfterburnerPara[7] + (fNLAfterburnerPara[8] * newEnergy));
+        }
+
+        //correction only works in a specific momentum range
+        if (oldEnergy > 0.300) {
+        	  clus->SetNonLinCorrEnergy(newEnergy);
+          if (fSetForceClusterE) clus->SetE(newEnergy);
+        }
+        else {
+      	  	  AliWarning(Form("Cluster energy out of range for correction!, E = %f \n",oldEnergy));
+      	  		clus->SetNonLinCorrEnergy(-100);
+        	  if (fSetForceClusterE) clus->SetE(-100);
+        }
+      }
+      else {
+        	//No correction has been determined for this data set yet - switch this component off for this period
+        	newEnergy = oldEnergy;
+      	  clus->SetNonLinCorrEnergy(newEnergy);
+      	  if (fSetForceClusterE) clus->SetE(newEnergy);
+      }
+
+      // Fill histograms only if cluster is not exotic, as in ClusterMaker (the clusters are flagged, not removed)
+      if (fCreateHisto && !clus->GetIsExotic()) {
+        Float_t energy = clus->GetNonLinCorrEnergy();
+        if (fSetForceClusterE) energy = clus->E();
+        
+        fEnergyDistAfter->Fill(energy);
+        fEnergyTimeHistAfter->Fill(energy, clus->GetTOF());
+      }
+    }
+  }
+  return kTRUE;
+}
+
+// returns the period enumerator for a given period string
+// can be used to obtain the enum for GetCorrectedEnergy
+//________________________________________________________________________
+TString AliEmcalCorrectionClusterNonLinearityMCAfterburner::SummarizeMCProductions(TString namePeriod)
```

Comment:
I don't think I have the right answer yet, and I know that you've kept this the similar because we're going to have parallel tasks (which I think is less than ideal, but I guess is out of our control), but I see the current structure of these functions as being very difficult to maintain.

Taking a very quick look, it looks like we can't just make this function static in the original class and only maintain one copy due to dependency issues and type compatibility (please correct me if I'm wrong). Since it looks like these functions are already modified from the original class, I think we should modify them to improve the data structures (use a map, YAML (maybe not so good here), something). Perhaps we can try to mirror where things are defined so it's a bit easier to transfer values from one place to the other?

As I said, I don't think I have the right answers here - I'll need to think about it some more. @jdmulligan - thoughts?